### PR TITLE
use std::rotate

### PR DESCRIPTION
**Pull request by qndel**

### CODE REVIEW

File: Source/engine/palette.cpp [hunk 1]
```diff
@@ -203,7 +173,7 @@ void ApplyGamma(SDL_Color *dst, const SDL_Color *src, int n)
 void palette_init()
 {
 	LoadGamma();
-	memcpy(system_palette, orig_palette, sizeof(orig_palette));
+	memcpy(system_palette.data(), orig_palette.data(), sizeof(orig_palette));
```
Comment (glebm) [inline Source/engine/palette.cpp:176]:
```suggestion
	system_palette = orig_palette;
```

File: Source/engine/palette.cpp [hunk 2]
```diff
@@ -351,7 +321,7 @@ void PaletteFadeIn(int fr)
 		RenderPresent();
 	}
 
-	memcpy(logical_palette, orig_palette, sizeof(orig_palette));
+	memcpy(logical_palette.data(), orig_palette.data(), orig_palette.size());
```
Comment (glebm) [inline Source/engine/palette.cpp:324]:
```suggestion
	logical_palette = orig_palette;
```

File: Source/engine.cpp [hunk 3]
```diff
@@ -26,7 +26,7 @@ namespace {
 void DrawHalfTransparentUnalignedBlendedRectTo(const Surface &out, unsigned sx, unsigned sy, unsigned width, unsigned height)
 {
 	uint8_t *pix = out.at(static_cast<int>(sx), static_cast<int>(sy));
-	const uint8_t *lookupTable = paletteTransparencyLookup[0];
+	const auto &lookupTable = paletteTransparencyLookup[0];
```
Comment (glebm) [inline Source/engine.cpp:29]:
```suggestion
	const std::array<uint8_t, 256> &lookupTable = paletteTransparencyLookup[0];
```

File: Source/engine/render/blit_impl.hpp [hunk 4]
```diff
@@ -72,7 +72,7 @@ DVL_ALWAYS_INLINE DVL_ATTRIBUTE_HOT void BlitFillBlended(uint8_t *dst, unsigned
 {
 	assert(length != 0);
 	const uint8_t *end = dst + length;
-	const uint8_t *tbl = paletteTransparencyLookup[color];
+	const auto &tbl = paletteTransparencyLookup[color];
```
Comment (qndel) [inline Source/engine/render/blit_impl.hpp:75]:
```suggestion
	const std::array<uint8_t, 256> &tbl = paletteTransparencyLookup[color];
```

File: Source/engine/palette.cpp [hunk 5]
```diff
@@ -150,28 +135,13 @@ void CycleColors(int from, int to)
...
 void CycleColorsReverse(int from, int to)
 {
-	{
-		SDL_Color col = system_palette[to];
-		for (int i = to; i > from; i--) {
-			system_palette[i] = system_palette[i - 1];
-		}
-		system_palette[from] = col;
-	}
+	std::rotate(system_palette.begin() + from + 1, system_palette.begin() + from, system_palette.begin() + to);
```
Comment (AJenbo) [inline Source/engine/palette.cpp:138]: Is this actually cycling them in the right to left direction?

Comment (qndel) [reply Source/engine/palette.cpp:138]: yes

**Pull request merged**